### PR TITLE
travis: run tests on macOS

### LINKED ISSUE #3888: doctests regression with v3.7.3 on OS-X
- [x] Include a detailed description of the bug or suggestion

After upgrading to v3.7.3, all doctests started failing with the exception:

```
<my-pyenv>/lib/python3.7/site-packages/_pytest/doctest.py:206: TypeError: cannot unpack non-iterable NoneType object
```

Everything is OK with v3.7.2.

Slightly longer traceback (For the record, looking into the stacktrace, I see some references to "darwin"-specific functions):

```
___________________________________________________________________ [doctest] xxx ___________________________________________________________________

self = <CallInfo when='call' exception: cannot unpack non-iterable NoneType object>, func = <function call_runtest_hook.<locals>.<lambda> at 0x1112f49d8>, when = 'call'
treat_keyboard_interrupt_as_exception = False

    def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        self.when = when
        self.start = time()
        try:
>           self.result = func()

<my-pyenv>/lib/python3.7/site-packages/_pytest/runner.py:201:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   lambda: ihook(item=item, **kwds),
    when=when,
    treat_keyboard_interrupt_as_exception=item.config.getvalue("usepdb"),
        )

<my-pyenv>/lib/python3.7/site-packages/_pytest/runner.py:183:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_HookCaller 'pytest_runtest_call'>, args = (), kwargs = {'item': <DoctestItem 'somemodule.somefunction'>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.argnames:
            notincall = set(self.argnames) - set(['__multicall__']) - set(
                kwargs.keys())
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call"
                    .format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)

<my-pyenv>/lib/python3.7/site-packages/pluggy/hooks.py:258:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.config.PytestPluginManager object at 0x1034317b8>, hook = <_HookCaller 'pytest_runtest_call'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '<my-pyenv......t 0x103d4f4e0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x10e8f9860>>]
kwargs = {'item': <DoctestItem 'somemodule.somefunction'>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

<my-pyenv>/lib/python3.7/site-packages/pluggy/manager.py:67:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

hook = <_HookCaller 'pytest_runtest_call'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '<my-pyenv......t 0x103d4f4e0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x10e8f9860>>]
kwargs = {'item': <DoctestItem 'somemodule.somefunction'>}

    self._inner_hookexec = lambda hook, methods, kwargs: \
        hook.multicall(
            methods, kwargs,
>           firstresult=hook.spec_opts.get('firstresult'),
        )

<my-pyenv>/lib/python3.7/site-packages/pluggy/manager.py:61:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

hook_impls = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '<my-pyenv......t 0x103d4f4e0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x10e8f9860>>]
caller_kwargs = {'item': <DoctestItem 'somemodule.somefunction'>}, firstresult = False

    def _multicall(hook_impls, caller_kwargs, firstresult=False):
        """Execute a call into multiple python functions/methods and return the
        result(s).

        ``caller_kwargs`` comes from _HookCaller.__call__().
        """
        __tracebackhide__ = True
        results = []
        excinfo = None
        try:  # run impl and wrapper setup functions in a loop
            teardowns = []
            try:
                for hook_impl in reversed(hook_impls):
                    try:
                        args = [caller_kwargs[argname] for argname in hook_impl.argnames]
                    except KeyError:
                        for argname in hook_impl.argnames:
                            if argname not in caller_kwargs:
                                raise HookCallError(
                                    "hook call must provide argument %r" % (argname,))

                    if hook_impl.hookwrapper:
                        try:
                            gen = hook_impl.function(*args)
                            next(gen)   # first yield
                            teardowns.append(gen)
                        except StopIteration:
                            _raise_wrapfail(gen, "did not yield")
                    else:
                        res = hook_impl.function(*args)
                        if res is not None:
                            results.append(res)
                            if firstresult:  # halt further impl calls
                                break
            except BaseException:
                excinfo = sys.exc_info()
        finally:
            if firstresult:  # first result hooks return a single value
                outcome = _Result(results[0] if results else None, excinfo)
            else:
                outcome = _Result(results, excinfo)

            # run all wrapper post-yield blocks
            for gen in reversed(teardowns):
                try:
                    gen.send(outcome)
                    _raise_wrapfail(gen, "has second yield")
                except StopIteration:
                    pass

>           return outcome.get_result()

<my-pyenv>/lib/python3.7/site-packages/pluggy/callers.py:201:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pluggy.callers._Result object at 0x11116efd0>

    def get_result(self):
        """Get the result(s) for this hook call.

            If the hook was marked as a ``firstresult`` only a single value
            will be returned otherwise a list of results.
            """
        __tracebackhide__ = True
        if self._excinfo is None:
            return self._result
        else:
            ex = self._excinfo
            if _py3:
>               raise ex[1].with_traceback(ex[2])

<my-pyenv>/lib/python3.7/site-packages/pluggy/callers.py:76:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

hook_impls = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '<my-pyenv......t 0x103d4f4e0>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x10e8f9860>>]
caller_kwargs = {'item': <DoctestItem 'somemodule.somefunction'>}, firstresult = False

    def _multicall(hook_impls, caller_kwargs, firstresult=False):
        """Execute a call into multiple python functions/methods and return the
        result(s).

        ``caller_kwargs`` comes from _HookCaller.__call__().
        """
        __tracebackhide__ = True
        results = []
        excinfo = None
        try:  # run impl and wrapper setup functions in a loop
            teardowns = []
            try:
                for hook_impl in reversed(hook_impls):
                    try:
                        args = [caller_kwargs[argname] for argname in hook_impl.argnames]
                    except KeyError:
                        for argname in hook_impl.argnames:
                            if argname not in caller_kwargs:
                                raise HookCallError(
                                    "hook call must provide argument %r" % (argname,))

                    if hook_impl.hookwrapper:
                        try:
                            gen = hook_impl.function(*args)
                            next(gen)   # first yield
                            teardowns.append(gen)
                        except StopIteration:
                            _raise_wrapfail(gen, "did not yield")
                    else:
>                       res = hook_impl.function(*args)

<my-pyenv>/lib/python3.7/site-packages/pluggy/callers.py:180:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

item = <DoctestItem 'somemodule.somefunction'>

    def pytest_runtest_call(item):
        _update_current_test_var(item, "call")
        sys.last_type, sys.last_value, sys.last_traceback = (None, None, None)
        try:
>           item.runtest()

<my-pyenv>/lib/python3.7/site-packages/_pytest/runner.py:111:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <DoctestItem 'somemodule.somefunction'>

    def runtest(self):
        _check_all_skipped(self.dtest)
>       self._disable_output_capturing_for_darwin()

<my-pyenv>/lib/python3.7/site-packages/_pytest/doctest.py:192:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <DoctestItem 'somemodule.somefunction'>

    def _disable_output_capturing_for_darwin(self):
        """
            Disable output capturing. Otherwise, stdout is lost to doctest (#985)
            """
        if platform.system() != "Darwin":
            return
        capman = self.config.pluginmanager.getplugin("capturemanager")
        if capman:
>           out, err = capman.suspend_global_capture(in_=True)
E           TypeError: cannot unpack non-iterable NoneType object

<my-pyenv>/lib/python3.7/site-packages/_pytest/doctest.py:206: TypeError
```

- [x] `pip list` of the virtual environment you are using

```
...
mypy                     0.620
pytest                   3.7.3
pytest-cov               2.5.1
pytest-django            3.4.2
tox                      3.2.1
...
```

- [x] pytest and operating system versions

`pytest v3.7.3` and 

```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G65
```

- [ ] Minimal example if possible

### LINKED ISSUE #3892: Test with OS-X on Travis
As #3888 shows, we have specific code which can be broken if we don't test for it.

### PR DESCRIPTION
**Pull request by jirikuncar**
Closes #3892 
Closes #3888

Thanks for submitting a PR, your contribution is really appreciated!

Here's a quick checklist that should be present in PRs (you can delete this text from the final description, this is
just a guideline):

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](/changelog/README.rst) for details.
- [x] Target the `master` branch for bug fixes, documentation updates and trivial changes.
- [ ] Target the `features` branch for new features and removals/deprecations.
- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.

Unless your change is trivial or a small documentation fix (e.g.,  a typo or reword of a small section) please:

- [ ] ~Add yourself to `AUTHORS` in alphabetical order;~ trivial change - no code

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -44,6 +44,16 @@ jobs:
       python: '3.7'
       sudo: required
       dist: xenial
+    - language: generic
```
Comment (nicoddemus) [inline .travis.yml]: Awesome. 😎 

Is `python2.7` available on this image? If so, is it worth it to also test on 2.7?

**Comment by nicoddemus** [comment]
@jirikuncar the build failed with:

```
/Users/travis/.travis/job_stages: line 78: tox: command not found
```

Can you install `tox` with brew?

Btw, feel free to comment out the other environments until we get OS-X running, so the builds are faster. 👍

File: .travis.yml [hunk 2]
```diff
@@ -44,6 +44,25 @@ jobs:
       python: '3.7'
       sudo: required
       dist: xenial
+    - &test-macos
+      language: generic
+      os: osx
+      osx_image: xcode9.4
+      sudo: required
+      env: TOXENV=py27
```
Comment (blueyed) [inline .travis.yml]: Should not be defined in the "included" part probably ("&test-macos").

File: .travis.yml [hunk 3]
```diff
@@ -44,6 +44,25 @@ jobs:
...
+      install:
+        - python -m pip install --upgrade --pre tox
+    - <<: *test-macos
+      env: TOXENV=py36
+      install:
+        - python3 -m pip install --upgrade --pre tox
```
Comment (blueyed) [inline .travis.yml]: Doesn't the normal "install" suffice?  (also above)

File: .travis.yml (hunk 2 above)
Comment (nicoddemus) [reply .travis.yml]: done

**Comment by nicoddemus** [comment]
@jirikuncar I force-pushed to your branch to remove a temporary commit, hope that's OK.

I've fixed the failing test and dropped py36 from OS-X: in my experience Travis OS-X jobs take longer (sometimes a lot longer) to start than plain linux jobs, and I think testing with py27 and py37 is enough to ensure compatibility. 👍

**Comment by jni** [comment]
Thanks everyone for the fast response here! Can I ask whether a patch release will be made soon (e.g. ~1 week timeframe)? We need to decide whether to pin to an earlier pytest or wait for this fix in our own repo.

File: .travis.yml (hunk 2 above)
Comment (jirikuncar) [reply .travis.yml]: It's good to have a default which get changed later.

**Comment by nicoddemus** [comment]
@jni we strive to make bug fix releases every 2 weeks or so, but we can make an exception and cut a release sooner, no problem. But regardless I would suggest to pin pytest to an earlier version, otherwise your PRs will be broken until then.

**Comment by jirikuncar** [comment]
@jni I have personally changed my dependency to `pytest>=3.6.3,!=3.7.3` so I get the latest update when it is ready.

**Comment by jni** [comment]
Thanks everyone! Yes this episode has taught me that `!=` is a valid pip limiter. =) Which is what I was after, something that would not pull in 3.7.3 but would not prevent future updates.

@nicoddemus the reason I was asking is that our project is pretty slow moving at the moment and I can't reliably predict that we will merge a fix before 3.7.4 is out. =P

**Pull request merged**